index.js

```diff
@@ -87,13 +91,14 @@ const matcher = (inputs, patterns, options, firstMatchOnly) => {
 		if (matches || (matches === undefined && !patterns.some(pattern => !pattern.negated))) {
 			result.push(input);
 
-			if (firstMatchOnly) {
+			if (firstMatchOnly && !allPatterns) {
 				break;
 			}
 		}
 	}
 
-	return result;
+	return (!allPatterns ||
+		didFit.every((flag, i) => flag ^ (patterns[i].negated & 1))) ? result : [];
```

Comment:
```suggestion
		didFit.every((flag, index) => flag ^ (patterns[index].negated & 1))) ? result : [];
```

Reply:
Well... I like the new code better, thanks for the hint!